libcudacxx/docs/extended_api/memory_resource.md

```diff
@@ -0,0 +1,220 @@
+struct my_memory_resource {
+    void* allocate(std::size_t, std::size_t) { return nullptr; }
+    void deallocate(void*, std::size_t, std::size_t) {}
+    bool operator==(const my_memory_resource&) const { return true; }
+    bool operator!=(const my_memory_resource&) const { return false; }
+
+    friend constexpr std::size_t get_property(const my_memory_resource& resource, required_alignment) noexcept { return resource.required_alignment; }
+
+    std::size_t required_alignment;
+};
+
+template<class MemoryResource>
+    requires cuda::mr::resource<MemoryResource>
+void* maybe_allocate_async_check_alignment(MemoryResource& resource, std::size_t size, cuda::stream_ref stream) {
+    if constexpr(cuda::mr::async_resource_with<MemoryResource, desired_alignment>) {
+        return resource.allocate_async(size, get_property(resource, required_alignment), stream);
+    } else if constexpr (cuda::mr::async_resource<MemoryResource>) {
+        return resource.allocate_async(size, my_default_alignment, stream);
+    } else if constexpr (cuda::mr::resource_with<MemoryResource, desired_alignment>) {
+        return resource.allocate(size, get_property(resource, required_alignment));
+    } else {
+        return resource.allocate(size, my_default_alignment);
+    }
+}
+
+// Potentially more consice
+template<class MemoryResource>
+    requires cuda::mr::resource<MemoryResource>
+void* maybe_allocate_async_check_alignment2(MemoryResource& resource, std::size_t size, cuda::stream_ref stream) {
+    constexpr std::size_t align = cuda::mr::resource_with<MemoryResource, desired_alignment>
+                                ? get_property(resource, required_alignment)
+                                : my_default_alignment;
+    if constexpr(cuda::mr::async_resource<MemoryResource>) {
+        return resource.allocate_async(size, align, stream);
+    } else {
+        return resource.allocate(size, align);
+    }
+}
+```
+
+### `resource_ref` a type constrained resource wrapper
+
+With the property design shown above a library has maximal flexibility regarding constraints checks and querying for any custom property. However, there is also a cost in providing function templates for a potentially wide range of inputs. Depending on the number of different memory resources both compile times and binary size might increase considerably.
+
+To efficiently coalesce such APIs into a single function we provide the type erased `resource_ref` and `async_resource_ref` resource wrappers.
+```c++
+void* do_allocate_async(cuda::mr::async_resource_ref<> resource, std::size_t size, std::size_t align, cuda::stream_ref stream) {
+    return resource.allocate_async(size, align, stream);
+}
+
+my_async_memory_resource resource;
+my_async_memory_resource* pointer_to_resource = &resource;
+
+void* from_reference = do_allocate_async(resource, 1337, 256, cuda::stream_ref{});
+void* from_ptr = do_allocate_async(pointer_to_resource, 1337, 256, cuda::stream_ref{});
+
```

Comment:
```suggestion
```

Reply:
I believe the comment did not come through